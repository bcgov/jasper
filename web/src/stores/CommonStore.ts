import {
  ApplicationInfo,
  CourtRoomsJsonInfoType,
  IconStyleType,
  LookupCode,
  UserInfo,
} from '@/types/common';
import { defineStore } from 'pinia';

enum appearanceStatus {
  UNCF = 'Unconfirmed',
  CNCL = 'Canceled',
  SCHD = 'Scheduled',
}

export const useCommonStore = defineStore('CommonStore', {
  state: () => ({
    displayName: '',
    userInfo: null as UserInfo | null,
    appInfo: null as ApplicationInfo | null,
    time: '',
    duration: '',
    statusStyle: '',
    iconStyles: [] as IconStyleType[],
    courtRoomsAndLocations: [] as CourtRoomsJsonInfoType[],
    roles: [] as LookupCode[],
    enableArchive: false,
    email: '',
  }),
  actions: {
    setUserInfo(userInfo): void {
      this.userInfo = userInfo;
    },
    updateUserInfo(newUserInfo): void {
      this.setUserInfo(newUserInfo);
    },
    setCourtRoomsAndLocations(courtRoomsAndLocations): void {
      this.courtRoomsAndLocations = courtRoomsAndLocations;
    },
    updateCourtRoomsAndLocations(newCourtRoomsAndLocations) {
      this.setCourtRoomsAndLocations(newCourtRoomsAndLocations);
    },
    setRoles(roles): void {
      this.roles = roles;
    },
    updateRoles(newRoles): void {
      this.setRoles(newRoles);
    },
    setDisplayName(displayName): void {
      this.displayName = displayName;
    },
    updateDisplayName(inputNames): void {
      let newDisplayName = '';
      if (inputNames.lastName.length == 0) {
        newDisplayName = inputNames.givenName;
      } else if (inputNames.givenName.length == 0) {
        newDisplayName = inputNames.lastName;
      } else {
        newDisplayName =
          inputNames.lastName.charAt(0).toUpperCase() +
          inputNames.lastName.slice(1).toLowerCase() +
          ', ' +
          inputNames.givenName.charAt(0).toUpperCase() +
          inputNames.givenName.slice(1).toLowerCase();
      }
      this.setDisplayName(newDisplayName);
    },
    formatDisplayName(newDisplayName): void {
      this.setDisplayName(
        newDisplayName.charAt(0).toUpperCase() +
          newDisplayName.slice(1).toLowerCase()
      );
    },
    setTime(time): void {
      this.time = time;
    },
    updateTime(time) {
      const time12 = (Number(time.substr(0, 2)) % 12 || 12) + time.substr(2, 3);

      if (Number(time.substr(0, 2)) < 12) {
        this.setTime(time12 + ' AM');
      } else {
        this.setTime(time12 + ' PM');
      }
    },
    setDuration(duration): void {
      this.duration = duration;
    },
    updateDuration(newDuration) {
      let duration = '';
      if (newDuration.hr) {
        if (Number(newDuration.hr) == 1) duration += '1 Hr ';
        else if (Number(newDuration.hr) > 1)
          duration += Number(newDuration.hr) + ' Hrs ';
      }

      if (newDuration.min) {
        if (Number(newDuration.min) == 1) duration += '1 Min ';
        else if (Number(newDuration.min) > 1)
          duration += Number(newDuration.min) + ' Mins ';
      }

      this.setDuration(duration);
    },
    setStatusStyle(statusStyle): void {
      this.statusStyle = statusStyle;
    },
    updateStatusStyle(status) {
      let style = '';
      if (status == appearanceStatus.UNCF) {
        style = 'badge badge-danger mt-2';
      } else if (status == appearanceStatus.CNCL) {
        style = 'badge badge-warning mt-2';
      } else if (status == appearanceStatus.SCHD) {
        style = 'badge badge-primary mt-2';
      }

      this.setStatusStyle(style);
    },
    setIconStyle(iconStyles): void {
      this.iconStyles = iconStyles;
    },
    updateIconStyle(newIconsInfo) {
      const iconStyles: IconStyleType[] = [];
      for (const iconInfo of newIconsInfo) {
        if (iconInfo['info'] == 'UNCF') {
          iconStyles.push({ icon: 'circle-half', desc: appearanceStatus.UNCF });
        } else if (iconInfo['info'] == 'CNCL') {
          iconStyles.push({ icon: 'trash', desc: appearanceStatus.CNCL });
        } else if (iconInfo['info'] == 'SCHD') {
          iconStyles.push({ icon: 'calendar', desc: appearanceStatus.SCHD });
        } else if (iconInfo['info'] == 'Video') {
          iconStyles.push({ icon: 'camera-video-fill', desc: 'video' });
        } else if (iconInfo['info'] == 'Home') {
          iconStyles.push({ icon: 'house-door-fill', desc: iconInfo['desc'] });
        }
      }
      this.setIconStyle(iconStyles);
    },
    setEnableArchive(newEnableArchive): void {
      this.enableArchive = newEnableArchive;
    },
    UpdateEnableArchive(newEnableArchive): void {
      this.setEnableArchive(newEnableArchive);
    },
  },
});

export type CommonStore = ReturnType<typeof useCommonStore>;
