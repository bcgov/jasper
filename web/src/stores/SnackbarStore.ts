import { defineStore } from 'pinia';
import { ref } from 'vue';

export const useSnackbarStore = defineStore('snackbar', () => {
  const isVisible = ref(false);
  const message = ref('');
  const color = ref('success');
  const title = ref('');
  const timeout = ref<number>();

  const showSnackbar = (msg = '', col = 'success', ti = '', time = 15000) => {
    message.value = msg;
    color.value = col;
    title.value = ti;
    isVisible.value = true;
    timeout.value = time;
  };

  const hideSnackbar = () => {
    isVisible.value = false;
  };

  return {
    isVisible,
    message,
    color,
    showSnackbar,
    hideSnackbar,
    title,
    timeout,
  };
});
