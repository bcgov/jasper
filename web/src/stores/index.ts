import { createPinia } from 'pinia';
import { App } from 'vue';
import piniaPluginPersistedstate from 'pinia-plugin-persistedstate'

const pinia = createPinia();
pinia.use(piniaPluginPersistedstate)

export function registerPinia(app: App) {
  app.use(pinia);
}

export default pinia;

export { useCivilFileStore } from './CivilFileStore';
export { useCommonStore } from './CommonStore';
export { useCourtFileSearchStore } from './CourtFileSearchStore';
export { useCourtListStore } from './CourtListStore';
export { useCriminalFileStore } from './CriminalFileStore';
export { useSnackbarStore } from './SnackbarStore';
export { usePDFViewerStore } from './PDFViewerStore';
export { useBundleStore } from './BundleStore';