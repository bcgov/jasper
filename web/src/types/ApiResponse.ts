export interface ApiResponse<T> {
  payload: T;
  succeeded: boolean;
  errors: string[];
}
