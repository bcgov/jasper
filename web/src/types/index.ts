export * from './Activity';
export * from './AdjudicatorRestriction';
export * from './Binder';
export * from './BinderDocument';
export * from './CalendarDay';
export * from './CalendarDayActivity';
export * from './CourtCalendarSchedule';
export * from './JudicialCalendarActivity';
export * from './JudicialCalendarAssignment';
export * from './Location';
export * from './PersonSearchItem';
export * from './Presider';
export * from './Tag';
