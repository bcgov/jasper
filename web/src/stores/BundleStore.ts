import { defineStore } from 'pinia';
import { UUIDTypes } from 'uuid';
import { Binder } from '@/types/Binder';
  import {
    CourtListAppearanceDocumentRequest,
    CourtListDocumentBundleRequest,
  } from '@/types/courtlist/jsonTypes';


export const useBundleStore = defineStore('BundleStore', {
  persist: true,
  state: () => ({
    bundles: [] as Bundle[],
    request: {} as CourtListDocumentBundleRequest,
    appearanceRequests: [] as appearanceRequest[],
  }),
  getters: {
    getBundle: (state): ((id: UUIDTypes) => Bundle | undefined) => {
      return (id: UUIDTypes) => {
        const bundle = state.bundles.find(b => b.id === id);
        return bundle;
      };
    },
    getRequests: (state) => state.request,
    getAppearanceRequests: (state) => state.appearanceRequests,
  },
  actions: {
    addBundle(id: UUIDTypes): void {
      this.bundles.push({
        id: id,
        binders: [] as Binder[],
        groupKeyOne: '',
        groupKeyTwo: '',
        documentName: '',
        requests: {} as CourtListDocumentBundleRequest,
      });
    },
    addBinder(binder: Binder, bundleId: UUIDTypes): void {
      const bundle = this.bundles.find(b => b.id === bundleId);
      if (bundle) {
        bundle.binders.push(binder);
      }
    },
    clearBundles(): void {
      this.bundles.length = 0;
      this.appearanceRequests = [];
      this.request = {} as CourtListDocumentBundleRequest;
    },
  },
});

export type Bundle = {
  id: UUIDTypes;
  groupKeyOne: string;
  groupKeyTwo: string;
  documentName: string;
  requests: CourtListDocumentBundleRequest;
  binders: Binder[];
};

export type appearanceRequest = {
  appearance: CourtListAppearanceDocumentRequest;
  fileNumber: string;
  fullName: string;
};