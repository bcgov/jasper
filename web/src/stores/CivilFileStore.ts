import { civilFileInformationType } from '@/types/civil';
import { defineStore } from 'pinia';

export const useCivilFileStore = defineStore('CivilFileStore', {
  state: () => ({
    civilFileInformation: {} as civilFileInformationType,
    civilFileInfoLoaded: false,
    hasNonParty: false,
    showSections: {
      'Case Details': true,
      'Future Appearances': false,
      'Past Appearances': false,
      'All Documents': false,
      Documents: false,
      'Provided Documents': false,
    },
    civilAppearanceInfo: {
      supplementalEquipmentTxt: '',
      securityRestrictionTxt: '',
      outOfTownJudgeTxt: '',
      fileNo: '',
      appearanceId: '',
      date: '',
    },
  }),
  actions: {
    setCivilFile(civilFileInformation): void {
      this.civilFileInformation = civilFileInformation;
    },
    updateCivilFile(newCivilFileInformation): void {
      this.setCivilFile(newCivilFileInformation);
    },
    setCivilFileInfoLoaded(civilFileInfoLoaded): void {
      this.civilFileInfoLoaded = civilFileInfoLoaded;
    },
    updateCivilFileInfoLoaded(newCivilFileInfoLoaded): void {
      this.setCivilFileInfoLoaded(newCivilFileInfoLoaded);
    },
    setHasNonParty(hasNonParty: boolean): void {
      this.hasNonParty = hasNonParty;
    },
    updateHasNonParty(newHasNonParty: boolean): void {
      this.setHasNonParty(newHasNonParty);
    },
    setShowSections(showSections): void {
      this.showSections = showSections;
    },
    updateShowSections(newShowSections): void {
      this.setShowSections(newShowSections);
    },
    setCivilAppearanceInfo(civilAppearanceInfo): void {
      this.civilAppearanceInfo = civilAppearanceInfo;
    },
    updateCivilAppearanceInfo(newCivilAppearanceInfo): void {
      this.setCivilAppearanceInfo(newCivilAppearanceInfo);
    },
  },
});
