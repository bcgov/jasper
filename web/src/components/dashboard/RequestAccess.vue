<template>
  <v-container class="home" fluid>
    <v-row justify="center">
      <v-col cols="12" md="8" lg="4">
        <v-card
          class="text-center access-body"
          elevation="2"
          style="padding: 1rem"
        >
          <img src="../../assets/jasper.png" alt="B.C. Government Logo" />
          <v-row justify="center">
            <h2 class="mt-4">Request Access To JASPER:</h2>
          </v-row>
          <v-row justify="center">
            <v-col cols="12" sm="10" md="8" xs="10">
              <v-text-field
                id="email"
                v-model="selectedEmail"
                label="Email"
                :disabled="true"
                persistent-hint
                hint="This email is associated to the account you logged in with, it will be used to create your account."
                type="email"
                class="email-input"
              >
              </v-text-field>
            </v-col>
          </v-row>
          <v-row justify="center">
            <v-btn
              color="primary"
              class="py-2 px-5"
              @click="requestAccess"
              :disabled="isSubmitted || isUserInvalid || isUserDisabled"
            >
              {{ isSubmitted ? 'Submitted' : 'Submit Your Request' }}
            </v-btn>
          </v-row>
          <v-row justify="center" class="my-4">
            <v-col cols="12" xs="12">
              <v-alert
                v-if="isSubmitted"
                border="start"
                type="success"
                class="shared-badge px-5 mx-auto"
              >
                Your request has been submitted!<br />
                We will get back to you soon.
              </v-alert>
              <v-alert
                v-if="isUserDisabled"
                border="start"
                type="warning"
                class="shared-badge px-5 mx-auto"
              >
                Your user has been disabled.<br />
                Please contact the JASPER admin if you require access.
              </v-alert>
              <v-alert
                v-if="isUserInvalid"
                border="start"
                type="warning"
                class="shared-badge px-5 mx-auto"
              >
                Warning, you do not have valid access to JASPER.<br />
                Please contact the JASPER admin to correct your account.
              </v-alert>
            </v-col>
          </v-row>
        </v-card>
      </v-col>
    </v-row>
  </v-container>
</template>

<script setup lang="ts">
  import { UserService } from '@/services/UserService';
  import { useCommonStore } from '@/stores';
  import { useSnackbarStore } from '@/stores/SnackbarStore';
  import { CustomAPIError, isPositiveInteger } from '@/utils/utils';
  import axios, { AxiosError } from 'axios';
  import _ from 'underscore';
  import { inject, onMounted, ref } from 'vue';

  const commonStore = useCommonStore();
  const snackBarStore = useSnackbarStore();
  const selectedEmail = ref<string>(commonStore.userInfo?.email ?? '');
  const isUserDisabled = ref(false);
  const isUserInvalid = ref(false);
  const isSubmitted = ref(false);
  const userService = inject<UserService>('userService');

  const requestAccess = async (): Promise<void> => {
    if (!_.isEmpty(selectedEmail.value)) {
      try {
        const accessRequest = await userService?.requestAccess(
          selectedEmail.value
        );
        if (accessRequest?.email === selectedEmail.value) {
          isSubmitted.value = true;
        } else {
          throw Error();
        }
      } catch (error) {
        if (
          error instanceof CustomAPIError &&
          axios.isAxiosError(
            (error as CustomAPIError<AxiosError>).originalError
          )
        ) {
          snackBarStore.showSnackbar(
            `${(error as CustomAPIError<AxiosError<[]>>)?.originalError?.response?.data?.join(' ') ?? ''}`,
            '#b84157',
            'Unable to submit access request'
          );
        }
      }
    }
  };

  const checkForUser = async (): Promise<void> => {
    const user = await userService?.getMyUser();

    if (!user?.isActive && isPositiveInteger(user?.roles?.length)) {
      isUserDisabled.value = true;
    } else if (user?.isPendingRegistration) {
      isSubmitted.value = true;
    } else if (
      (user?.isActive && !isPositiveInteger(user?.roles?.length)) ||
      !user
    ) {
      isUserInvalid.value = true;
    }
  };

  onMounted(async () => {
    await checkForUser();
  });
</script>

<style scoped>
  .card {
    border: white;
  }

  .btn.disabled {
    cursor: default;
  }

  .shared-badge {
    font-size: 16px;
    font-weight: unset;
    white-space: wrap;
  }

  .access-body {
    background-color: var(--bg-blue-100);
  }

  .email-input :deep() input {
    text-align: center;
  }
</style>
