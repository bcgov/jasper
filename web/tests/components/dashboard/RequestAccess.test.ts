import RequestAccess from '@/components/dashboard/RequestAccess.vue';
import { useCommonStore } from '@/stores';
import { useSnackbarStore } from '@/stores/SnackbarStore';
import { CustomAPIError } from '@/types/ApiResponse';
import { flushPromises, mount } from '@vue/test-utils';
import { AxiosError } from 'axios';
import { createPinia, setActivePinia } from 'pinia';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import { nextTick } from 'vue';

// Mock UserService
const mockRequestAccess = vi.fn();
const mockGetMyUser = vi.fn();
const mockUserService = {
  requestAccess: mockRequestAccess,
  getMyUser: mockGetMyUser,
};

// Mock stores
vi.mock('@/stores');
vi.mock('@/services');

// Helper to mount component with provide
const mountComponent = async (options = {}) => {
  const wrapper = mount(RequestAccess, {
    global: {
      provide: {
        userService: mockUserService,
      },
    },
    ...options,
  });
  await flushPromises();
  return wrapper;
};

describe('RequestAccess.vue', () => {
  let snackStore: ReturnType<typeof useSnackbarStore>;
  beforeEach(() => {
    setActivePinia(createPinia());
    (useCommonStore as any).mockReturnValue({
      userInfo: { email: 'test@example.com' },
    });
    snackStore = useSnackbarStore();
  });

  it('renders correctly (snapshot)', async () => {
    mockGetMyUser.mockResolvedValue(undefined);
    const wrapper = await mountComponent();

    await vi.waitFor(() => {
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  it('disables input and button when isSubmitted is true', async () => {
    mockGetMyUser.mockResolvedValue({
      email: 'test@example.com',
      isPendingRegistration: true,
    });

    const wrapper = await mountComponent();

    expect(wrapper.find('v-text-field').attributes('disabled')).toBeDefined();
    expect(wrapper.find('v-btn').attributes('disabled')).toBeDefined();
    expect(wrapper.text()).toContain('Your request has been submitted!');
  });

  it('shows user disabled badge if user is inactive with roles', async () => {
    mockGetMyUser.mockResolvedValue({
      email: 'test@example.com',
      isActive: false,
      roles: [1, 2],
    });

    const wrapper = await mountComponent();
    await nextTick();

    expect(wrapper.text()).toContain('Your user has been disabled.');
  });

  it('shows user invalid badge if user is not pending, not disabled', async () => {
    mockGetMyUser.mockResolvedValue({
      email: 'test@example.com',
      isActive: true,
      isPendingRegistration: false,
      roles: [],
    });

    const wrapper = await mountComponent();
    await nextTick();

    expect(wrapper.text()).toContain(
      'Warning, you do not have valid access to JASPER.'
    );
  });

  it('calls requestAccess and sets isSubmitted on success', async () => {
    mockGetMyUser.mockResolvedValue(undefined);
    mockRequestAccess.mockResolvedValue({ email: 'test@example.com' });

    const wrapper = await mountComponent();
    await nextTick();
    wrapper.find('v-btn').trigger('click');
    await flushPromises();

    expect(mockRequestAccess).toHaveBeenCalledWith('test@example.com');
    expect(wrapper.text()).toContain('Your request has been submitted!');
  });

  it('calls requestAccess and displays errors returned from backend', async () => {
    mockGetMyUser.mockResolvedValue(undefined);
    const axiosError = new AxiosError('test error');
    mockRequestAccess.mockRejectedValue(
      new CustomAPIError(axiosError.message, axiosError)
    );
    snackStore.showSnackbar = vi.fn();

    const wrapper = await mountComponent();
    await nextTick();
    wrapper.find('v-btn').trigger('click');
    await nextTick();

    expect(mockRequestAccess).toHaveBeenCalledWith('test@example.com');
    expect(snackStore.showSnackbar).toHaveBeenCalled();
  });

  it('does not call requestAccess if email is empty', async () => {
    (useCommonStore as any).mockReturnValue({
      userInfo: { email: '' },
    });
    mockGetMyUser.mockResolvedValue(undefined);

    const wrapper = mount(RequestAccess, {
      global: { provide: { userService: mockUserService } },
    });
    await flushPromises();
    wrapper.find('v-btn').trigger('click');
    await nextTick();

    expect(mockRequestAccess).not.toHaveBeenCalled();
  });
});
