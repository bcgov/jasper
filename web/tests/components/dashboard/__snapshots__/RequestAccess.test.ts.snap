// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RequestAccess.vue > renders correctly (snapshot) 1`] = `
<v-container
  class="home"
  fluid=""
>
  <v-row justify="center">
    <v-col
      cols="12"
      lg="4"
      md="8"
    >
      <v-card
        class="
          access-body
          text-center
        "
        elevation="2"
        style="padding: 1rem;"
      >
        <img
          alt="B.C. Government Logo"
          src="/src/assets/jasper.png"
        />
        <v-row justify="center">
          <h2 class="mt-4">
            Request Access To JASPER:
          </h2>
        </v-row>
        <v-row justify="center">
          <v-col
            cols="12"
            md="8"
            sm="10"
            xs="10"
          >
            <v-text-field
              class="email-input"
              disabled="true"
              hint="This email is associated to the account you logged in with, it will be used to create your account."
              id="email"
              label="Email"
              modelvalue="test@example.com"
              persistent-hint=""
              type="email"
            ></v-text-field>
          </v-col>
        </v-row>
        <v-row justify="center">
          <v-btn
            class="
              px-5
              py-2
            "
            color="primary"
            disabled="true"
          >
            Submit Your Request
          </v-btn>
        </v-row>
        <v-row
          class="my-4"
          justify="center"
        >
          <v-col
            cols="12"
            xs="12"
          >
            <!-- v-if -->
            <!-- v-if -->
            <v-alert
              border="start"
              class="
                mx-auto
                px-5
                shared-badge
              "
              type="warning"
            >
              Warning, you do not have valid access to JASPER.
              <br />
              Please contact the JASPER admin to correct your account.
            </v-alert>
          </v-col>
        </v-row>
      </v-card>
    </v-col>
  </v-row>
</v-container>
`;
